### PR TITLE
Add more checks to linter. Fixes #959 and #890.

### LINKED ISSUE #890: Add remaining schema Style Guide checks to schema linter
**Description**

 

As a metadata team member (committer) who makes changes to the metadata schemas, I would like to be sure my schema changes adhere to the Style Guide by running the schema linter.

Features are from this [doc](https://docs.google.com/document/d/1pg-YCBE9YGAEVVZQN6p3nKMghdjiUSd8i4mlnRheoUs/edit#).

**Acceptance Criteria**

 
The following checks should be added to the linter:
 
- [x] Every schema title should be sentence-case (first char is uppercase letter, no underscores).
- [x] Every field that references another schema using $ref must point to a valid schema.
- [x] For every field that has a pattern attribute, the pattern must be a valid regular expression.
- [x] Example values for fields with a pattern must match the regular expression.
- [x] ~For every field that is of type array, the field name should be plural.~ not worth the effort to implement right now
- [x] Fields must be allowed to have `oneOf` as a property.

Ontology-specific checks to be added to the linter:

- [x] Each ontology schema must have text, ontology, and ontology_label fields
- [x] Every ontology field must have a graph_restriction attribute that is an object
- [x] Every graph_restriction must have an ontologies attribute
    - [x] ~Must be valid curies~ Moved to issue https://github.com/HumanCellAtlas/metadata-schema/issues/1012
    - [x] ~Check that each listed ontology is present in HCA OLS instance~ Moved to issue https://github.com/HumanCellAtlas/metadata-schema/issues/1012
- [x] Every graph_restriction must have a classes attribute
    - [x] ~Must be valid curies~ Moved to issue https://github.com/HumanCellAtlas/metadata-schema/issues/1012
    - [x] ~Check that each listed class is present in HCA OLS instance~ Moved to issue https://github.com/HumanCellAtlas/metadata-schema/issues/1012
- [x] Every graph_restriction must have a relations attribute
    - [x] Relation attribute is a list that must minimally be ["rdfs:subClassOf"]
- [x] Every graph_restriction must have a direct attribute
    - [x] Must be false
- [x] Every graph_restriction must have a include_self attribute
    - [x] Must be true or false

### LINKED ISSUE #959: Schema linter shows Errors and Warnings separately. Fails if any Errors.
**Description**

 

As a DCP member who updates the metadata schemas, I would like to know whether the updates I did result in deal-breaking violations of the Style Guide (errors that need to be fixed before committing) or result in undesirable but acceptable violations of the Style Guide (warnings that _should_ be fixed but are OK to commit).

**Acceptance Criteria**

  
- [x] Output of running the schema linter: passes if no warnings or errors, passes if warnings but no errors, and fails if any errors. 
- [x] Warnings and errors describe what the violation is.

### PR DESCRIPTION
**Pull request by malloryfreeberg**
### Release notes

- No schema changes.
- Added more checks to linter. Fixes #890.
- Added failures if "musts" are violated, warnings otherwise. Fixes #959.

### Reviews requested

- Need 1 Reviewers to approve because this is a code update.
- Last Reviewer to review/approve, merge changes by **Tuesday June 11**

### CODE REVIEW

File: src/schema_linter.py [hunk 1]
```diff
@@ -26,139 +39,213 @@ def __init__(self):
...
+                        # Attributes for graph_restriction must be one of acceptable values
+                        if nkw not in ontology_attributes:
+                            sys.exit(schema_filename + ".json: Keyword `" + nkw + "` is not an acceptable graph_restriction keyword property.")
+
+                    # graph_restriction 'direct' attribute must be 'false'
+                    if properties['ontology']['graph_restriction']['direct'] is not False:
+                        sys.exit(schema_filename + ".json: Keyword 'direct' must be set to 'false', not " + str(properties['ontology']['graph_restriction']['direct']) + ".")
+
+                    # graph_restriction 'include_self' attribute must be 'false' or 'true'
+                    if properties['ontology']['graph_restriction']['include_self'] not in [False, True]:
```
Comment (ESapenaVentura) [inline src/schema_linter.py]: Just to maintain consistency, I would use `if not isinstance(properties['ontology']['graph_restriction']['include_self'], bool):`

File: src/schema_linter.py [hunk 2]
```diff
@@ -26,139 +39,213 @@ def __init__(self):
...
-            print(schema_filename + ".json: The `type` attribute must be set to object")
+            sys.exit(schema_filename + ".json: The `type` attribute must be set to \"object\".")
 
-        # Check that all required fields are actually in the schema
+        # All required fields must actually be in the schema
         if "required" in schema:
             for req_prop in schema["required"]:
                 if req_prop not in properties:
-                    print("Property `" + req_prop + "` is required in " + schema_filename + ".json but is undefined")
+                    sys.exit("Property `" + req_prop + "` is required in " + schema_filename + ".json but is missing under properties.")
```
Comment (ESapenaVentura) [inline src/schema_linter.py]: Same with the exit_flag. If you end up using it, you might even define it before the class.

File: src/schema_linter.py [hunk 3]
```diff
@@ -26,139 +39,213 @@ def __init__(self):
...
 
-        schema_filename = path.split("/")[-1].split(".")[0]
-
-        # Check that all root level fields in the schema are part of the list of allowed root level fields
+        # All schema fields must be part of the list of allowed schema fields
         for key in schema.keys():
-            if key not in allowed_root_level_keywords:
-                print("Root level field `" + key + "` in schema " + path + " not part of allowed root level properties")
+            if key not in allowed_schema_fields:
+                sys.exit("Schema field `" + key + "` in " + schema_filename + ".json" + " not an allowed schema field.")
```
[the remark this answers is not in the document]
Comment (malloryfreeberg) [reply src/schema_linter.py]: That does makes sense. It's probably better to catch/print all the fail errors and fail once at the end, rather then ending at the first fail error. So there will just be 1 `sys.exit()` at the end. Do you want to go ahead and implement your suggestion?

File: src/schema_linter.py (hunk 1 above)
Comment (malloryfreeberg) [reply src/schema_linter.py]: Good idea. Will make this change

**Comment by malloryfreeberg** [comment]
@ESapenaVentura I just made a few updates to the linter. Please pull if you've already started implementing your updates!

**Comment by ESapenaVentura** [comment]
Looks good!

**Pull request merged**